### PR TITLE
Update model.cpp load_3ds to support Big Endian

### PR DESCRIPTION
**Pull request by rt1k**
Fix 3ds model loader for big endian architectures; tested on PowerPC 64 Debian bullseye/sid.

### CODE REVIEW

File: src/model.cpp [hunk 1]
```diff
@@ -32,6 +32,7 @@
 #include <fstream>
 #include <sstream>
 #include <memory>
+#include <byteswap.h>
```
Comment (afrantzis) [inline src/model.cpp:35]: The byteswap.h header does not exist on platform we support (e.g., Windows). We need something more portable, which means either providing manually implemented swaps, or conditionally including platform specific headers.

File: src/model.cpp [hunk 2]
```diff
@@ -49,6 +50,25 @@ using LibMatrix::uvec3;
     } \
 } while(0);
 
+const int isBigEnd=1;
+#define is_bigendian() ((*(char*)&isBigEnd) == 0)
+
+float FloatSwap( float f )
```
Comment (afrantzis) [inline src/model.cpp:56]: No space after/before parentheses, i.e., `float FloatSwap(float f)`

File: src/model.cpp [hunk 3]
```diff
@@ -373,10 +393,16 @@ Model::load_3ds(const std::string &filename)
     while (!input_file.eof()) {
         uint16_t chunk_id;
         uint32_t chunk_length;
-
+		uint32_t swap32t;
```
Comment (afrantzis) [inline src/model.cpp:396]: glmark2 uses 4 spaces for indentation (not tabs). Also applies to all code below.

File: src/model.cpp [hunk 4]
```diff
@@ -373,10 +393,16 @@ Model::load_3ds(const std::string &filename)
...
+		uint16_t swap16t;
+		
         // Read the chunk header
         input_file.read(reinterpret_cast<char *>(&chunk_id), 2);
-        if (input_file.gcount() == 0) {
+		if(is_bigendian()) {
+			swap16t=__bswap_16(chunk_id); chunk_id=swap16t;
```
Comment (afrantzis) [inline src/model.cpp:402]: The byteswap.h API is `bswap_16/32/64()`. The underscored versions are internal implementation details.

Also, AFAIK, there is no need for an intermediate variable. We could more simply write: `chunk_id = is_bigendian() ? bswap_16(chunk_id) : chunk_id;` (like it's already done for some cases below).

Similarly for all swaps below.

File: src/model.cpp [hunk 5]
```diff
@@ -49,6 +50,25 @@ using LibMatrix::uvec3;
     } \
 } while(0);
 
+const int isBigEnd=1;
+#define is_bigendian() ((*(char*)&isBigEnd) == 0)
+
+float FloatSwap( float f )
+{
+  union
```
Comment (afrantzis) [inline src/model.cpp:58]: glmark2 uses 4 spaces for indentation (not 2 spaces).

**Review by afrantzis — CHANGES_REQUESTED** [review]
Thanks for PR! I have a few comments inline.

File: src/model.cpp (hunk 1 above)
Comment (rt1k) [reply src/model.cpp]: How could i test it on windows ? there are building instructions to build on windows? thanks

File: src/model.cpp [hunk 6]
```diff
@@ -49,6 +70,25 @@ using LibMatrix::uvec3;
...
+float FloatSwap(float f)
+{
+    union
+    {
+        float f;
+        unsigned char b[4];
+    } dat1, dat2;
+
+    dat1.f = f;
+    dat2.b[0] = dat1.b[3];
```
Comment (afrantzis) [inline src/model.cpp:85]: It turns out that, per the C++ standard, it's undefined behavior to read from a union member that wasn't explicitly written to before (although most compilers seem to deal with it OK). We can safely cast a float to char*, though, and achieve the same result:

```
float FloatSwap(float f)
{
    float ret;
    char *fbytes = reinterpret_cast<char*>(&f);
    char *retbytes = reinterpret_cast<char*>(&ret);
   ...
}

**Comment by afrantzis** [comment]
Hi and thanks again for the PR! I have decided to merge the alternative proposal from #152, so we can close this now.

**Pull request closed without merging**